### PR TITLE
Add Fog Of War 

### LINKED ISSUE #3602: Add fog of war for the patch map
only show the patches next to where the player's species is or the player has been before. We can probably make it persistent so that once some patch has become visible it stays permanently visible. Regions with no visible patches should probably be hidden. Visible regions should probably have their connection lines to other regions visible to give a hint where there are more regions. Freebuild should ignore the fog of war.

### PR DESCRIPTION
**Pull request by ender-titan1**
**Brief Description of What This PR Does**

This PR is the second version of my implementation of Fog Of War to the patch map. This is a much nicer and more elegant solution to Fog Of War, with most of the code governing patch visibility contained in `PatchMap` and `Patch` itself.

Link to my previous PR (Just in case): #4462 

Feature Checklist:

- [X] Saving & Loading
- [x] Reveal All Cheat
- [x] Fading Region Connections
- [x] Making Connections To Unexplored Regions Look Nice
- [x] AutoEvo Output & Report 
- [x] Translations 

**Related Issues**

Closes #3602 

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by ender-titan1** [comment]
The previous commit is supposed to make connections to unknown regions look nicer, which was what held up the development of my previous PR. This time, I did not bother changing the way connections are rendered, instead just centering a single node in the region. 

This yields ok results, albeit still a bit disconnected from the rest of the map. I think this is fine for now, and trying to make it look nicer should be done by someone more experienced with the patch map code, and probably be a separate issue. 

I believe the results of this are good enough for fog-of-war itself.

**Comment by ender-titan1** [comment]
This is now ready for review

File: src/microbe_stage/editor/MicrobeEditorPatchMap.tscn [hunk 1]
```diff
@@ -115,7 +115,6 @@ margin_left = 10.0
 margin_top = 10.0
 margin_right = 896.0
 margin_bottom = 608.0
-ShowScrollbars = false
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditorPatchMap.tscn:118]: What's this change about?

File: src/gui_common/fonts/Thrive-Regular-Humongous.tres [hunk 2]
```diff
@@ -13,8 +13,9 @@
 [ext_resource path="res://assets/fonts/NotoSansBengali-SemiBold.ttf" type="DynamicFontData" id=11]
 
 [resource]
-size = 65
+size = 32
```
Comment (hhyyrylainen) [inline src/gui_common/fonts/Thrive-Regular-Humongous.tres:16]: Is this font used anywhere else? I'd really prefer a new font size to be made. I wouldn't call 32 size even humongous anymore...

File: src/microbe_stage/editor/MicrobeEditorPatchMap.tscn (hunk 1 above)
Comment (ender-titan1) [reply src/microbe_stage/editor/MicrobeEditorPatchMap.tscn:118]: Looks like Godot is automatically changing things again... I'll revert this

File: src/microbe_stage/editor/PatchMapDrawer.cs [hunk 3]
```diff
@@ -734,14 +757,15 @@ private void DrawPatchLinks()
...
         foreach (var node in nodes.Values)
-        {
             node.Free();
-        }
 
         nodes.Clear();
+
+        RebuildRegionConnections();
+
         connections.Clear();
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/PatchMapDrawer.cs:769]: I have just one thing I can't wrap my head around: why is this called here and then immediately the connections list is cleared? If the connection list rebuilt in _Draw ends up with different data, that data is then never used by RebuildRegionConnections except on the first render when the player patch is focused.

**Review by hhyyrylainen — APPROVED** [review]
I've fixed all of the slight code issues I saw and fixed up comment typos etc. I've also tested this quickly for each fog of war mode and also loading one earlier save. I think this is now good to go by me, but I'll wait a bit until merging to give others time to double check my changes that I didn't accidentally break anything.

File: src/microbe_stage/editor/PatchMapDrawer.cs (hunk 3 above)
Comment (ender-titan1) [reply src/microbe_stage/editor/PatchMapDrawer.cs:772]: I've fixed this by calling it after connections have been rebuilt

**Comment by hhyyrylainen** [comment]
I think you undid my squashing of this PR... I think I'll be able to fix it tomorrow by cherry picking the following commit: `3777ee2f42516501a7565c45af44edf0204bc21b` to my local branch and then force pushing over this again.

**Comment by ender-titan1** [comment]
Sorry, I'll keep this in mind for future PRs

**Comment by hhyyrylainen** [comment]
Alright, that was much easier than expected. Hopefully you didn't make any other changes than that one commit about where the region link build was. This should be fine to merge now with the cleaned up history.

**Comment by hhyyrylainen** [comment]
Seems like auto-evo exploring tool has broken map for all worlds after the first. I'll open an issue about that as a follow up to get this PR merged to no longer have to fix this branch: https://github.com/Revolutionary-Games/Thrive/issues/4731

**Pull request merged**